File: csharp/1255.maximum-score-words-formed-by-letters.cs
/*
 * @lc app=leetcode id=1255 lang=csharp
 *
 * [1255] Maximum Score Words Formed by Letters
 */

using System;

// @lc code=start
public class Solution {
    
    private static int MaxScoreWordsRecurse(string[] words, int[] score, int[] count, int start) {
        int ret = 0;
        for (int i = start; i < words.Length; ++i) {
            string curr = words[i];
            bool isOkay = true;
            int p = 0;
            foreach (char c in curr) {
                p += score[(int)c - (int)'a'];
                isOkay &= (--count[(int)c - (int)'a'] >= 0);
            }
            if (isOkay) {
                ret = Math.Max(ret, p + MaxScoreWordsRecurse(words, score, count, i + 1));
            }
            foreach (char c in curr) {
                ++count[(int)c - (int)'a'];
            }
        }
        return ret;
    }
    
    public int MaxScoreWords(string[] words, char[] letters, int[] score) {
        int[] count = new int[26];
        foreach (char c in letters) {
            ++count[(int)c - (int)'a'];
        }
        return MaxScoreWordsRecurse(words, score, count, 0);
    }
}
// @lc code=end

File: java/1255.maximum-score-words-formed-by-letters.java
/*
 * @lc app=leetcode id=1255 lang=java
 *
 * [1255] Maximum Score Words Formed by Letters
 */

// @lc code=start
class Solution {
    public int maxScoreWords(String[] words, char[] letters, int[] score) {
        int[] count = new int[26];
        for (char c : letters) {
            ++count[c - 'a'];
        }
        return helper(words, count, score, 0);
    }

    private int helper(String[] words, int[] count, int[] score, int curr) {
        int ret = 0;
        for (int i = curr; i < words.length; ++i) {
            String word = words[i];
            int len = word.length(), j, sc = 0;
            for (j = 0; j < len; ++j) {
                int c = word.charAt(j) - 'a';
                if (count[c] == 0) {
                    break;
                }
                sc += score[c];
                --count[c];
            }
            if (j == len) {
                ret = Math.max(ret, sc + helper(words, count, score, i + 1));
            }
            for (j--; j >= 0; --j) {
                ++count[word.charAt(j) - 'a'];
            }
        }
        return ret;
    }
}
// @lc code=end

